tools/run_oldest_tests.sh

```diff
@@ -0,0 +1,40 @@
+#!/bin/bash
+set -e
+
+DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+pushd "${DIR}/../" || exit 1
+
+function cleanup() {
+  rm -f "${SCRIPT}"
+  popd
+}
+
+trap cleanup EXIT
+
+SCRIPT=$(mktemp /tmp/test-script.XXXXXX)
+chmod +x "${SCRIPT}"
+
+cat << "EOF" >> "${SCRIPT}"
+#!/bin/bash
+set -e
+apt-get update
+apt-get install -y --no-install-recommends \
+    python-dev python-pip git gcc libaugeas0 libssl-dev libffi-dev \
+    ca-certificates nginx-light openssl curl software-properties-common
+apt-key adv --keyserver keyserver.ubuntu.com --recv-keys 5BB92C09DB82666C
+add-apt-repository ppa:fkrull/deadsnakes-python2.7
+apt-get update
+apt-get upgrade -y
+sh <(curl -fsSL https://get.docker.com)
+python -m pip install --upgrade pip virtualenv wheel
+python -m pip install tox
+python -m tox
+EOF
+
+docker run --rm --network=host -w "${PWD}" \
+  -v /var/run/docker.sock:/var/run/docker.sock \
+  -v "${PWD}:${PWD}" -v "${SCRIPT}:/script.sh" \
+  -v /tmp:/tmp \
+  -e TOXENV -e ACME_SERVER -e PYTEST_ADDOPTS \
+  ubuntu:14.04 /script.sh
```

Comment:
Regardless of the CI system we chose though, it seems we'll no longer be able to run on Ubuntu 14.04 outside of Docker.

Do you remember the problems you hit trying to run this test on Ubuntu 16.04? The version of `cryptography` we have pinned in our "oldest" tests is the same version as in Ubuntu 16.04 so I would expect it to build there unless the Ubuntu package fails to build from source or is customized in some way.

Reply:
Yes, you have very weird things happening with the specific set of oldest constraints. Like, on Ubuntu 16.04:
```
Collecting distlib<1,>=0.3.0 (from virtualenv)
  Downloading https://files.pythonhosted.org/packages/7d/29/694a3a4d7c0e1aef76092e9167fbe372e0f7da055f5dcf4e1313ec21d96a/distlib-0.3.0.zip (571kB)
    Complete output from command python setup.py egg_info:
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
    ImportError: No module named setuptools
```

In fact when you look at the log for the current oldest tests in Travis, you would notice a bunch of errors saying that a given dependency is installed at a specific version which is incompatible with another dependency. It just works "miraculously" with Ubuntu 14.04.

I also figured out that the "miracle" with Ubuntu 14.04 on Travis could be reproduced in other places (while I could not initially with anything other than Travis) if you install a more recent version of Python 2.7 that what is available in the official repositories (and so the use of `add-apt-repository ppa:fkrull/deadsnakes-python2.7` here).

Reply:
I got this to work outside of CI on Ubuntu 16.04 by updating `pyOpenSSL` to 0.15.1 to match the version packaged in Xenial. We could probably do this after checking in with Certbot's EPEL maintainers because while RHEL 7 still has `pyOpenSSL` 0.13.1, the Python 3 version packaged in EPEL is 17.3.0. We could probably just bump the dependency and they could relax it on their end to update the Certbot packages until they've finished the transition to Python 3.

This still doesn't work in Azure though because they've updated the OpenSSL version past what is normally in Ubuntu Xenial. Xenial usually has OpenSSL 1.0.2g but [Xenial on Azure has OpenSSL 1.1.0h](https://dev.azure.com/certbot/certbot/_build/results?buildId=2038&view=logs&j=a4a69b7d-afd1-5098-8e5e-d571ec92fd9a&t=7156f0f2-c52f-5c27-feb6-47d9b3991e2e&l=14).

Looking at [cryptography's changelog](https://github.com/pyca/cryptography/blob/master/CHANGELOG.rst#15---2016-08-26) it seemed like 1.5.0+ would be needed. To test this tried bumping `cryptography` to 1.7.1 which drops support for the version of cryptography in Xenial which we can also maybe do with more downsides, but [this doesn't work either for some reason](https://dev.azure.com/certbot/certbot/_build/results?buildId=2038&view=logs&j=a4a69b7d-afd1-5098-8e5e-d571ec92fd9a).

If you feel like digging into this more, you're welcome to do so, but otherwise, I think it's probably best to rely on Docker. This doesn't necessarily have to be done in this PR, but I could imagine us updating the "oldest" test setup in `tox.ini` so Docker is always used allowing people to run the test locally if they want.

Reply:
Yes the oldest tests requires a very specific crafted system, on Ubuntu 14.04 or another distribution. Docker is certainly the good tool here to ensure repeatability.

What was your reason to move the oldest tests on 16.04? Not relying on Docker to execute them?

Also, it is certainly outside of the scope of this PR, but given the numerous adaptations needed to run the oldest tests to run on a specific system, I wonder if the oldest tests still represents something. Even with 14.04 you still need to update to a recent Python 2.7 which is not obvious at all to install (and honestly I do not know how much the deadsnackes PPA will stay, it is an archived project).

Given that Python 2.7 will be removed in Certbot soon, shouldn't we rethink totally the oldest tests with Python 3 in mind?

Reply:
Again, I don't think it needs to be that specially crafted as the tests work on a vanilla Ubuntu 16.04 system after updating `PyOpenSSL` slightly, but you do need to compile things against an ancient version of OpenSSL which can get tricky.

> What was your reason to move the oldest tests on 16.04? Not relying on Docker to execute them?

Not relying on Docker was my initial motivation, but always relying on Docker so devs can run this locally seems like a good idea. My motivation now is to stop using the EOL'd Ubuntu 14.04 and the deadsnakes PPA.

> Given that Python 2.7 will be removed in Certbot soon, shouldn't we rethink totally the oldest tests with Python 3 in mind?

I started to write out a long response to this, but instead for now I'll say I'm curious what you have in mind, but I don't think we should worry about that in this PR :)

Reply:
I asked Certbot's Fedora/EPEL maintainer about this at https://github.com/certbot/certbot/issues/8110.

Reply:
Our Fedora/EPEL maintainer said us updating the PyOpenSSL dependency should be fine. I think this should be very easy to do so we may as well do it now.

I think we should:

1. Update `pyOpenSSL` in our oldest constraints to 0.15.1 and move it under the section for Xenial.
2. Update the minimum version of `pyOpenSSL` in `acme` (and maybe add a minimum version to Certbot and the nginx plugin).
3. Update the code here to use Ubuntu Xenial rather than Trusty.

Bonus points for setting the use of Docker in `tox`, but if that takes significant time, I think it's OK for us to do it later.

Reply:
I did the migration to Xenial for the oldest tests.

However concerning delegating the call to `tox`, I am not convinced at this time. Indeed the script `tools/run_oldest_test.py` is designed as a "wrapper" of the tox call, which executes it in a proper oldest environment, and so all the tox goodies (venv setup, TOXENV analysis and so on) are done directly in it. If we invert the logic, then `tools/run_oldest_test.py` will need to take care about it, and so will add some complexity.

Anyway adding this is not trivial, so I think we should live it outside this PR.